presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java

```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
             throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Concatenated string is too large");
         }
     }
+
+    private static class ConcatSignatureBinder
+            extends SignatureBinder
+    {
+        public ConcatSignatureBinder(TypeManager typeManager, Signature signature, boolean allowCoercion)
+        {
+            super(typeManager, signature, allowCoercion);
+        }
+
+        @Override
+        public Optional<Signature> bind(List<? extends Type> actualArgumentTypes)
+        {
+            ImmutableList.Builder<TypeSignature> boundTypesSignatures = ImmutableList.builder();
+
+            long lengthAccumulator = 0;
+            for (Type type : actualArgumentTypes) {
+                if (!isTypeValid(type)) {
+                    return Optional.empty();
+                }
+
+                int paramLength = varcharLength(type);
+                lengthAccumulator = Ints.saturatedCast(lengthAccumulator + paramLength);
```

Comment:
Becasue max length of VARCHAR is MAX_INT and two ints may have bigger value. In such case we just want to retrurn VARCHAR(MAX_INT).

Typical case for that is when one of CONCAT arguments has length of MAX_INT as it is result of other calculation not allowing to limit VARCHAR size to anything smaller.